### PR TITLE
fix prng key reuse in differential privacy example

### PR DESCRIPTION
**Pull request by mattjj**
cc @lberrada @BorjaBalle

This was a pretty insidious bug: the differential privacy example was accidentally reusing the same PRNG key when adding noise to different components of the gradient (i.e. corresponding to different parameters). Since JAX currently leaves PRNG key reuse totally unchecked, there were no guardrails to catch this!

(It was mitigated somewhat by another peculiarity: sampling random noise for different shapes (perhaps different ranks) produces distinct entries, even with the same key. But the right fix is not to reuse PRNG keys, and someday for us to check for key reuse.)

### CODE REVIEW

**Comment by lberrada** [comment]
Thanks for the responsiveness, happy to take a look.

File: examples/differentially_private_sgd.py [hunk 1]
```diff
@@ -134,33 +133,30 @@ def accuracy(params, batch):
...
-    nonempty_grads, tree_def = tree_util.tree_flatten(grads)
-    total_grad_norm = jnp.linalg.norm(
-        [jnp.linalg.norm(neg.ravel()) for neg in nonempty_grads])
-    divisor = stop_gradient(jnp.amax((total_grad_norm / l2_norm_clip, 1.)))
-    normalized_nonempty_grads = [g / divisor for g in nonempty_grads]
-    return tree_util.tree_unflatten(tree_def, normalized_nonempty_grads)
-
-  px_clipped_grad_fn = vmap(partial(_clipped_grad, params))
-  std_dev = l2_norm_clip * noise_multiplier
-  noise_ = lambda n: n + std_dev * random.normal(rng, n.shape)
```
Comment (BorjaBalle) [inline examples/differentially_private_sgd.py:154]: Fixing `rng` in `noise_` will introduce dependencies on the noise added across layers in a single update.
The Gaussian mechanism in differential privacy requires independently sampled noise in each parameter.

Comment (mattjj) [reply examples/differentially_private_sgd.py:154]: Indeed! (This is a good diagnosis of the original problem this PR is fixing, since this comment is on deleted code.)

File: examples/differentially_private_sgd.py [hunk 2]
```diff
@@ -134,33 +132,30 @@ def accuracy(params, batch):
...
-  tree_map = tree_util.tree_map
-  sum_ = lambda n: jnp.sum(n, 0)  # aggregate
-  aggregated_clipped_grads = tree_map(sum_, px_clipped_grad_fn(batch))
-  noised_aggregated_clipped_grads = tree_map(noise_, aggregated_clipped_grads)
-  normalized_noised_aggregated_clipped_grads = (
-      tree_map(normalize_, noised_aggregated_clipped_grads)
-  )
-  return normalized_noised_aggregated_clipped_grads
+  clipped_grads = vmap(clipped_grad, (None, None, 0))(params, l2_norm_clip, batch)
+  clipped_grads_flat, params_tree = tree_flatten(clipped_grads)
```
Comment (lberrada) [inline examples/differentially_private_sgd.py:150]: nit / suggestion: `params_tree` could be renamed `grads_treedef` to make it clearer

**Comment by lberrada** [comment]
LGTM!

**Review by lberrada — APPROVED** [review]

**Review by BorjaBalle — APPROVED** [review]
LGTM, thanks for the quick response!

File: examples/differentially_private_sgd.py (hunk 2 above)
Comment (mattjj) [reply examples/differentially_private_sgd.py:150]: Good idea!

**Comment by mattjj** [comment]
Thanks for spotting this and pointing out the fix, @BorjaBalle @lberrada !

File: examples/differentially_private_sgd.py (hunk 2 above)
Comment (mattjj) [reply examples/differentially_private_sgd.py]: Done.
